tests/test_mem_eff_attention.py

```diff
@@ -48,49 +134,305 @@ def test_key_query_all_ones(device, q_len, kv_len, batch_size, k_len):
+        ("query", grad_q, query.grad),
+        ("key", grad_k, key.grad),
+        ("value", grad_v, value.grad),
+    ]:
+        assert_allclose(calc_grad, ref_grad, name, atol=atol, rtol=rtol)
+
+
+def _vec_binom_test(x, n, p):
+    """
+    vectorized implementation of scipy.stats.binom_test
+    this makes our tests much faster
+    reference: https://github.com/scipy/scipy/blob/v1.8.0/scipy/stats/_morestats.py#L2609-L2702
+    """
+    import numpy as np
+    from scipy.stats import distributions
+
+    x = np.atleast_1d(x)
+    d = distributions.binom.pmf(x, n, p)[:, None]
+    rerr = 1 + 1e-7
+    # x < p * n case
+    i = np.arange(np.ceil(p * n), n + 1)
+    y = np.sum(distributions.binom.pmf(i, n, p) <= d * rerr, axis=1)
+    pval1 = distributions.binom.cdf(x, n, p) + distributions.binom.sf(n - y, n, p)
+
+    # other case
+    i = np.arange(np.floor(p * n) + 1)
+    y = np.sum(distributions.binom.pmf(i, n, p) <= d * rerr, axis=1)
+    pval2 = distributions.binom.cdf(y - 1, n, p) + distributions.binom.sf(x - 1, n, p)
+
+    pval = np.where(x < p * n, pval1, pval2)
+    pval = np.minimum(1.0, pval)
+    return pval
+
+
+@cuda_only
+@pytest.mark.parametrize("seed", [42, 124])
+@pytest.mark.parametrize("p", [0.3, 0.7])
+@pytest.mark.parametrize("k_len", [32])
+@pytest.mark.parametrize("batch_size", [1, 4])
+@pytest.mark.parametrize("kv_len", [3, 15, 32, 33])
+@pytest.mark.parametrize("q_len", [2, 33])
+@pytest.mark.parametrize("device", ["cuda"])
+def test_dropout(device, q_len, kv_len, batch_size, k_len, p, seed):
     scale = 3
     query = torch.randn((batch_size, q_len, k_len), device=device) * scale
     key = torch.randn((batch_size, kv_len, k_len), device=device) * scale
     value = torch.randn((batch_size, kv_len, k_len), device=device) * scale
 
-    _, lse = torch.ops.xformers.efficient_attention(query, key, value, True)
-    ref_lse = ((query / k_len**0.5) @ key.transpose(-2, -1)).logsumexp(-1)
+    attn_bias = None
 
-    assert torch.allclose(lse, ref_lse, atol=2e-4)
+    torch.manual_seed(seed)
+    out = xformers.ops.memory_efficient_attention(query, key, value, attn_bias, p)
 
+    torch.manual_seed(seed)
+    out2 = xformers.ops.memory_efficient_attention(query, key, value, attn_bias, p)
 
-@pytest.mark.parametrize("grad_out_contiguous", [False, True])
+    assert_allclose(out, out2)
+
+    mask = torch.empty((batch_size, q_len, kv_len), device=device)
+
+    torch.manual_seed(seed)
+    mask = torch.ops.xformers._temp_dropout(mask, p)
+
+    ref = ref_attention(query, key, value, attn_bias, mask, p)
+    assert_allclose(out, ref, atol=2e-4), f"{(out - ref).abs().max()}"
+
+    num_trials = 1000
+    p_val_tol = 0.0001
+    keep_prob = 1 - p
+    masks = []
+    for i in range(num_trials):
+        mask = torch.ops.xformers._temp_dropout(mask, p)
+        masks.append(mask.clone().cpu())
+    masks = torch.stack(masks, dim=0)
+    p_value = binom_test(masks.sum(), masks.numel(), p=keep_prob)
```

Comment:
nice, seems much better than [these tests](https://github.com/facebookresearch/xformers/blob/main/tests/test_triton_dropout.py#L96) (my bad), thanks for this very thorough take